x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportSetUpgradeModeAction.java

```diff
@@ -0,0 +1,271 @@
+              assignment
+
+
+              We make no guarantees around which tasks will be running again once upgrade_mode is disabled.
+              Scenario:
+                 * Before `upgrade_mode: true`, there were unassigned tasks because node task assignment was maxed out (tasks A, B)
+                 * There were assigned tasks executing fine (tasks C, D)
+                 * While `upgrade_mode: true` all are attempting to be re-assigned, but cannot and end up with the AWAITING_UPGRADE reason
+                 * `upgrade_mode: false` opens the flood gates, all tasks are still attempting to re-assign
+                 * A or B could be re-assigned before either C or D. Thus, previously erred tasks are now executing fine, and previously
+                   executing tasks are now unassigned due to resource exhaustion.
+
+              We make no promises which tasks are executing if resources of the cluster are exhausted.
+          </.2>
+          </2>
+         */
+        ActionListener<AcknowledgedResponse> clusterStateUpdateListener = ActionListener.wrap(
+            acknowledgedResponse -> {
+                // State change was not acknowledged, we either timed out or ran into some exception
+                // We should not continue and alert failure to the end user
+                if (acknowledgedResponse.isAcknowledged() == false) {
+                    wrappedListener.onFailure(new ElasticsearchTimeoutException("Unknown error occurred while updating cluster state"));
+                    return;
+                }
+                // Did we change from disabled -> enabled?
+                if (request.isEnabled()) {
+                    isolateDatafeeds(tasksCustomMetaData, isolateDatafeedListener);
+                } else {
+                    persistentTasksService.waitForPersistentTasksCondition(
+                        (persistentTasksCustomMetaData) ->
+                            // Wait for jobs to not be "Awaiting upgrade"
+                            persistentTasksCustomMetaData.findTasks(JOB_TASK_NAME,
+                                (t) -> t.getAssignment().equals(AWAITING_UPGRADE))
+                                .isEmpty() &&
+
+                            // Datafeeds to wait for a non-"Awaiting upgrade" assignment and for the job task allocations to converge
+                            // If we do not wait, deleting datafeeds, or attempting to unallocate them again causes issues as the
+                            // job's task allocationId could have changed during either process.
+                            persistentTasksCustomMetaData.findTasks(DATAFEED_TASK_NAME,
+                                (t) ->
+                                    t.getAssignment().equals(AWAITING_UPGRADE) ||
+                                    t.getAssignment().getExplanation().contains("state is stale"))
```

Comment:
I think this "state is stale" condition is a sign that there will be situations where the endpoint could fail in production use.  A datafeed could be unassigned and have "state is stale" in its assignment reason if a node died shortly before `set_upgrade_mode?enabled=true` was called.  Presumably then this situation would have the same "issues" that calling `set_upgrade_mode?enabled=false` immediately followed by `set_upgrade_mode?enabled=true` has.

If this is a really hard problem to solve then we could merge this PR as-is so that the Kibana team have something to test against, but keep working on a followup to do whatever is required to remove this "state is stale" test.

Reply:
@droberts195 If the node died, the job would attempt to re-assign right?

* Could initiall fail due to resource constraints (assignmentId increments as Assignment reason changes)
* Then attempts again during upgrade (assignmentId increments as explanation reason is now due to the upgrade)
* Upgrade finishes, tries again and can either join a node or fail due to resource constraints, either way it should stabilize to a specific assignment? 

I am not sure who all updates that internal task state so that it would eventually come in sync with what the datafeed sees.